pkg/blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,509 @@
+		kgo.SeedBrokers(cfg.Address),
+		kgo.DialTimeout(cfg.DialTimeout),
+		kgo.MetadataMinAge(10 * time.Second),
+		kgo.MetadataMaxAge(10 * time.Second),
+		kgo.WithLogger(newKafkaLogger(logger)),
+	}
+	if metrics != nil {
+		opts = append(opts, kgo.WithHooks(metrics))
+	}
+	return opts
+}
+
+func (b *BlockBuilder) stopping(_ error) error {
+	b.kafkaClient.Close()
+
+	return nil
+}
+
+func (b *BlockBuilder) running(ctx context.Context) error {
+	// Do initial consumption on start using current time as the point up to which we are consuming.
+	// To avoid small blocks at startup, we consume until the last hour boundary + buffer.
+	cycleEnd := cycleEndAtStartup(b.cfg.ConsumeInterval, b.cfg.ConsumeIntervalBuffer)
+	err := b.NextConsumeCycle(ctx, cycleEnd)
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			return nil
+		}
+		return err
+	}
+
+	nextCycleTime := time.Now().Truncate(b.cfg.ConsumeInterval).Add(b.cfg.ConsumeInterval + b.cfg.ConsumeIntervalBuffer)
+	waitTime := time.Until(nextCycleTime)
+	for waitTime > b.cfg.ConsumeInterval {
+		// NB: at now=14:12, next cycle starts at 14:15 (startup cycle ended at 13:15)
+		//     at now=14:17, next cycle starts at 15:15 (startup cycle ended at 14:15)
+		nextCycleTime = nextCycleTime.Add(-b.cfg.ConsumeInterval)
+		waitTime -= b.cfg.ConsumeInterval
+	}
+
+	for {
+		select {
+		case <-time.After(waitTime):
+			cycleEnd := nextCycleTime
+			level.Info(b.logger).Log("msg", "triggering next consume cycle", "cycle_end", cycleEnd)
+			err := b.NextConsumeCycle(ctx, cycleEnd)
+			if err != nil {
+				b.metrics.consumeCycleFailures.Inc()
+				level.Error(b.logger).Log("msg", "consume cycle failed", "cycle_end", cycleEnd, "err", err)
+			}
+
+			// If we took more than ConsumeInterval to consume the records, this will immediately start the next consumption.
+			// TODO(codesome): track waitTime < 0, which is the time we ran over. Should have an alert on this.
+			nextCycleTime = nextCycleTime.Add(b.cfg.ConsumeInterval)
+			waitTime = time.Until(nextCycleTime)
+		case <-ctx.Done():
+			level.Info(b.logger).Log("msg", "context cancelled, stopping")
+			return nil
+		}
+	}
+}
+
+func cycleEndAtStartup(interval, buffer time.Duration) time.Time {
+	cycleEnd := time.Now().Truncate(interval).Add(buffer)
+	if cycleEnd.After(time.Now()) {
+		cycleEnd = cycleEnd.Add(-interval)
+	}
+	return cycleEnd
+}
+
+// NextConsumeCycle manages consumption of currently assigned partitions.
+// The cycleEnd argument indicates the timestamp (relative to Kafka records) up until which to consume from partitions
+// in this cycle. That is, Kafka records produced after the cycleEnd mark will be consumed in the next cycle.
+func (b *BlockBuilder) NextConsumeCycle(ctx context.Context, cycleEnd time.Time) error {
+	defer func(t time.Time) {
+		b.metrics.consumeCycleDuration.Observe(time.Since(t).Seconds())
+	}(time.Now())
+
+	for _, part := range b.parts {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if part < 0 {
```

Comment:
In which case could it happen? And... if happens, why do we stop iterating partitions and `return nil`?